### ISSUE TITLE
"Zoom" action shouldn't escape the "Pan" range bounds

### ISSUE THREAD

**Issue by ZaLiTHkA** [chartjs/chartjs-plugin-zoom#245]
I'm currently working to implement a chart with `chartjs-plugin-streaming` and `chartjs-plugin-zoom` to show real time data from a remote source, but I'm having a few issues with the zoom and pan side of my chart..

I've [created a JSFiddle](https://jsfiddle.net/ZaLiTHkA/0kf9ua42) inspired by the chart inside my app, showing the behaviour I'm having trouble with. to summarise the configuration:

* my app has a buffer with 5 minutes' worth of data
* this is shown in datasets linked to the `realtime` X-axis provided by the streaming plugin
* I have the zoom plugin's `pan.rangeMin` set to `0` and `pan.rangeMax` set to `5 * 60 * 1000` (matching the buffer duration)
* I'm *only* using pan and zoom on the X-axis, with Y-axis scaling controlled automatically by Chart.js

when I zoom my chart *out*, the chart zooms central to the point it was zoomed from (as expected), but if either the left or right bound of the "zoom" range extends past the bounds of the "pan" range, the chart retains that visible range.

at the point, trying to pan either left or right, only then is the pan location updated to fall within it's defined bounds.

I feel like this is an issue in the zooming plugin though, because I would expect the "zoom" action to stay within the "pan" bounds of the chart.

**Comment by ZaLiTHkA** [comment]
in the meantime, is there any way to programmatically trigger the "pan" action? I'm thinking I could perhaps trigger a "pan" action in both directions on the  X-axis in the `onZoom` callback, thereby forcing this plugin to execute the logic that checks if the visible area is still within the bounds set for the "pan" behaviour.

beyond that, I'd also like to be able to programmatically "pan" back to the very right end of the chart, which I'm guessing would actually be the same thing..

**Comment by benmccann** [comment]
Can you set `zoom.rangeMin = pan.rangeMin` and `zoom.rangeMax = pan.rangeMax`?

You chart sounds cool! I'd love to see it if it's publicly available to share!

**Comment by ZaLiTHkA** [comment]
> Can you set `zoom.rangeMin = pan.rangeMin` and `zoom.rangeMax = pan.rangeMax`?

I presume then if (for example) the value stored in `pan.rangeMax` updates, then `zoom.rangeMax` would also change.. very interesting idea.

I did have to put that chart aside for the moment to fix a few other bugs first, but I'll get back to it first thing in the morning (GMT+2), so I'll definitely try that and let you know what happens. :)

> You chart sounds cool! I'd love to see it if it's publicly available to share!

alas, this is part of a hybrid mobile app that communicates with proprietary traffic detectors, so you wouldn't really get any further than the landing page with a logo, an introduction, and an network address search field.

**Comment by ZaLiTHkA** [comment]
so I finally got a chance to try Ben's suggestion today (been swamped with other fixes and improvements lately), but unfortunately it didn't seem to work.

*however*, since the `rangeMin` and `rangeMax` options [behave differently](https://github.com/nagix/chartjs-plugin-streaming#support-for-zooming-and-panning) when this is combined with `chartjs-plugin-streaming`, should I perhaps ask this question over there instead?

or could the issue I'm seeing be resolved by simply ensuring that all "pan" limits are still applied after zooming?

---

there is one question I still haven't managed to find an answer for though: is it possible to programmatically control the "pan" position or "zoom" level with this plugin? for example, panning either the start or end of the visible area to a calculated point, which I could trigger from the `onZoom` callback?

**Comment by malle-pietje** [comment]
@ZaLiTHkA A bit late response but can't you do something like this which we use to sync the zoom/pan of two charts in the onZoom and onPan callbacks?
```javascript
chart_2.chart.options.scales.xAxes[0].time.min = chart_1.scales['x-axis-0'].min;
chart_2.chart.options.scales.xAxes[0].time.max = chart_1.scales['x-axis-0'].max;
```

**Comment by ZaLiTHkA** [comment]
so, [looks at the date] I finally got a chance to get back to this.. anyways, I got my desired behaviour working, my solution is two-fold:

1. I'm updating the pan "rangeMax" dynamically based on current chart data, in such a way that it prevents the "pan" action from moving past either end of the currently rendered chart data.
2. on panning or zooming, I clamp the chart's "delay" value between `0` and  a "sensible max delay" that I'm recalculating based on currently rendered chart data.

I implemented this in my hybrid mobile project and it's working pretty damn well, so I've updated [my original JSFiddle example](https://jsfiddle.net/ZaLiTHkA/0kf9ua42/218/) to share it back here again. :)
